dev/DynamicDependency/MddBootstrap.cpp

```diff
@@ -0,0 +1,378 @@
+#elif defined(_M_IX86)
+    return winrt::Windows::System::ProcessorArchitecture::X86;
+#elif defined(_M_ARM64)
+    return winrt::Windows::System::ProcessorArchitecture::Arm64;
+#elif defined(_M_ARM)
+    return winrt::Windows::System::ProcessorArchitecture::Arm;
+#else
+#   error "Unknown processor architecture"
+#endif
+}
+
+inline PCWSTR GetCurrentArchitectureAsString()
+{
+#if defined(_M_X64)
+    return L"x64";
+#elif defined(_M_IX86)
+    return L"x86";
+#elif defined(_M_ARM64)
+    return L"arm64";
+#elif defined(_M_ARM)
+    return L"arm";
+#else
+#   error "Unknown processor architecture"
+#endif
+}
+
+inline winrt::Windows::System::ProcessorArchitecture ParseArchitecture(PCWSTR architecture)
+{
+    if (CompareStringOrdinal(architecture, -1, L"x64", -1, TRUE) == CSTR_EQUAL)
+    {
+        return winrt::Windows::System::ProcessorArchitecture::X64;
+    }
+    else if (CompareStringOrdinal(architecture, -1, L"x86", -1, TRUE) == CSTR_EQUAL)
+    {
+        return winrt::Windows::System::ProcessorArchitecture::X86;
+    }
+    else if (CompareStringOrdinal(architecture, -1, L"arm64", -1, TRUE) == CSTR_EQUAL)
+    {
+    return winrt::Windows::System::ProcessorArchitecture::Arm64;
+    }
+    else if (CompareStringOrdinal(architecture, -1, L"arm", -1, TRUE) == CSTR_EQUAL)
+    {
+    return winrt::Windows::System::ProcessorArchitecture::Arm;
+    }
+    else
+    {
+        return winrt::Windows::System::ProcessorArchitecture::Unknown;
+    }
+}
+
+STDAPI MddBootstrapInitialize(
+    const PACKAGE_VERSION minVersion) noexcept try
+{
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_lifetimeManager != nullptr);
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_projectReunionDll != nullptr);
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_packageDependencyId != nullptr);
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_packageDependencyContext != nullptr);
+
+    const auto appDynamicDependencyLifetimeManagerClsid{ FindDDLM(minVersion) };
+
+    wil::com_ptr_nothrow<IDynamicDependencyLifetimeManager> lifetimeManager(wil::CoCreateInstance<IDynamicDependencyLifetimeManager>(appDynamicDependencyLifetimeManagerClsid, CLSCTX_LOCAL_SERVER));
+
+    THROW_IF_FAILED(lifetimeManager->Initialize());
+
+    wil::unique_cotaskmem_string packageFullName;
+    THROW_IF_FAILED(lifetimeManager->GetPackageFullName(&packageFullName));
+
+    const PACKAGE_INFO* frameworkPackageInfo{};
+    auto packageInfoBuffer{ GetFrameworkPackageInfoForPackage(packageFullName.get(), frameworkPackageInfo) };
+
+    // Temporarily add the framework's package directory to PATH so LoadLibrary can find it and any colocated imports
+    wil::unique_dll_directory_cookie dllDirectoryCookie{ AddFrameworkToPath(frameworkPackageInfo->path) };
+
+    auto projectReunionDllFilename{ std::wstring(frameworkPackageInfo->path) + L"\\Microsoft.ProjectReunion.dll" };
```

Comment:
At some point, we should look into making these DLL names parameterizable so the Mdd tooling can be used for other frameworks.

Reply:
Yep. There's several things we can genericize and reuse, in time. Some of the ground rules are a bit more wet-cement than graven-in-stone so I've been hesitant to run too far just yet.